src/main/groovy/com/github/winplay02/gitcraft/manifest/BaseMetadataProvider.java

```diff
@@ -0,0 +1,264 @@
+		this.manifestSources = new ArrayList<>();
+		this.metadataSources = new ArrayList<>();
+		this.repositorySources = new ArrayList<>();
+		this.loadSemverCache();
+	}
+
+	protected void addManifestSource(String url, Class<M> manifestClass) {
+		this.manifestSources.add(MetadataSources.RemoteVersionsManifest.of(url, manifestClass));
+	}
+
+	protected void addMetadataSource(E manifestEntry) {
+		this.metadataSources.add(MetadataSources.RemoteMetadata.of(manifestEntry));
+	}
+
+	protected void addMetadataRepository(String directory) {
+		this.repositorySources.add(MetadataSources.LocalRepository.of(this.localMetadata.resolve(directory)));
+	}
+
+	protected final Path getSemverCachePath() {
+		return LibraryPaths.CURRENT_WORKING_DIRECTORY.resolve(String.format("semver-cache-%s.json", this.getInternalName()));
+	}
+
+	protected final void loadSemverCache() {
+		Path cachePath = this.getSemverCachePath();
+		if (Files.exists(cachePath)) {
+			try {
+				this.semverCache.putAll(SerializationHelper.deserialize(SerializationHelper.fetchAllFromPath(cachePath), SerializationTypes.TYPE_TREE_MAP_STRING_STRING));
+			} catch (IOException e) {
+				MiscHelper.println("This is not a fatal error: %s", e);
+			}
+		}
+	}
+
+	protected final void writeSemverCache() {
+		Map<String, String> semverCache = new TreeMap<>(this.versionsById.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().semanticVersion())));
+		try {
+			SerializationHelper.writeAllToPath(this.getSemverCachePath(), SerializationHelper.serialize(semverCache));
+		} catch (IOException e) {
+			MiscHelper.println("This is not a fatal error: %s", e);
+		}
+	}
+
+	/**
+	 * @return A map containing all available versions, keyed by a unique name (see {@linkplain VersionInfo#id()}).
+	 */
+	@Override
+	public final Map<String, OrderedVersion> getVersions(Executor executor) throws IOException {
+		if (!this.versionsLoaded) {
+			this.loadVersions(executor);
+			this.postLoadVersions();
+			this.writeSemverCache();
+		}
+		return Collections.unmodifiableMap(this.versionsById);
+	}
+
+	private void loadVersions(Executor executor) throws IOException {
+		this.versionsById.clear();
+		MiscHelper.println("Loading available versions from '%s'...", this.getName());
+		for (MetadataSources.RemoteVersionsManifest<M, E> manifestSource : this.manifestSources) {
+			MiscHelper.println("Reading versions manifest from %s...", manifestSource.url());
+			M manifest = this.fetchVersionsManifest(manifestSource);
+			Map<String, CompletableFuture<OrderedVersion>> futureVersions = new HashMap<>();
+			for (E versionEntry : manifest.versions()) {
+				if (!this.versionsById.containsKey(versionEntry.id()) && !futureVersions.containsKey(versionEntry.id())) {
+					futureVersions.put(versionEntry.id(), this.loadVersionFromManifest(executor, versionEntry, this.manifestMetadata));
```

Comment:
after loading the version there should be a `!shouldExclude` check before adding it to the map